044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,173 @@
+* If more than one controller pod has not become ready, KafkaRoller should try to role the controllers that aren't the current active controller first.
+* If a broker pod has not become ready, KafkaRoller should check the controller quorum is formed before rolling the pod.
+* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 
+(i.e. not in pending state) before rolling the pod or waiting for it to become ready.
+
+The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
+The brokers do not move to RUNNING until a quorum has been formed.
+For example in combined mode during normal startup the following would happen:
+
+* All the pods are started at the same time
+* The controller processes in each pod form a quorum
+* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to STARTED and are marked as ready
+* All pods are now ready
+
+However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:
+
+* The first pod is scheduled
+* The controller process starts up but cannot form a quorum because the other controllers are missing
+* The broker process starts up but since there is no controller quorum it does not move to STARTED, it stays in STARTING. This means the pod is not marked as ready
+* Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff
+
+## Proposed order of tasks to complete this proposal
+
+Since changes to the KafkaRoller and support for non-combined mode is needed to fully implement this proposal, the following phases of development are proposed:
+
+### Phase 1 - Strimzi only supports combined mode
+* Combined mode pods are marked as both alive and ready when the pod is listening on 9090. 
+This is an improvement on the current state where the pod is marked as alive and ready as soon as it starts up.
+* Update the KafkaAgent to check for a broker state of >= 3 && != 127.
+This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.
+
+### Phase 2 - Strimzi adds support for broker only and controller only modes
+* Controller only pod liveness and readiness checks are fully implemented to match this proposal.
+* Broker only pods are marked as both alive and ready if listening on port 9091.
+* Combined mode pods are marked as both alive and ready when the pod is listening on 9090 (no change from phase 1).
+
+### Phase 3 - KafkaRoller is updated
+* KafkaRoller is updated to take the existence of controller pods and status of controller quorum into account when deciding whether to roll pods.
```

Comment:
I don't think there is any dependency between "supporting broker-only and controller-only mode" and "KafkaRoller updated", right? Maybe we should note that the phase 2 and phase 3 is not strictly in this order?

Reply:
I think there is a dependency. The KafkaRoller cannot be updated to understand the difference between broker and controller nodes until they are supported as part of Strimzi. These two phases could be done together in the same PR maybe but I don't think it makes sense to do phase 3 and then phase 2 as you would have to make guesses about how phase 2 would be implemented.

Reply:
Make sense. Thanks for the explanation.